### ISSUE TITLE
[Ubuntu 20.04] - Python 3.8 compatability causes few failed do_vigiles_pkg()

### ISSUE THREAD

**Issue by chaithubk** [TimesysGit/meta-timesys#1]
I happen to use the `Ubuntu 20.04 `as my Host machine to build the `core-image-sato` and ofcourse I get a Warning that says: 

`WARNING: Host distribution "ubuntu-20.04" has not been validated with this version of the build system; you may possibly experience unexpected failures. It is recommended that you use a tested distribution.`

But, still all the packages were built and even few `tsmeta` json files were created during this process. However there are few issues noticed after the `core-image-sato-1.0-r0 do_rootfs` with `do_vigiles_pkg` certain packages for example: 

```
ERROR: libinput-1.12.6-r0 do_vigiles_pkg: Error executing a python function in exec_python_func() autogenerated:

The stack trace of python calls that resulted in this exception/failure was:
File: 'exec_python_func() autogenerated', lineno: 2, function: <module>
     0001:
 *** 0002:do_tsmeta_pkgvars(d)
     0003:
File: '/home/poky/meta-timesys/classes/tsmeta.bbclass', lineno: 466, function: do_tsmeta_pkgvars
     0462:
     0463:
     0464:python do_tsmeta_pkgvars() {
     0465:    tsmeta_get_pn(d)
 *** 0466:    tsmeta_get_src(d)
     0467:    tsmeta_get_pkg(d)
     0468:    tsmeta_get_packageconfig(d)
     0469:}
     0470:
File: '/home/poky/meta-timesys/classes/tsmeta.bbclass', lineno: 302, function: tsmeta_get_src
     0298:    return cve_v
     0299:
     0300:
     0301:def tsmeta_get_src(d):
 *** 0302:    import oe.recipeutils as oe
     0303:
     0304:    tsm_type = "src"
     0305:    src_dict = dict()
     0306:
File: '/home/poky/meta/lib/oe/recipeutils.py', lineno: 21, function: <module>
     0017:import shutil
     0018:import re
     0019:import fnmatch
     0020:import glob
 *** 0021:import bb.tinfoil
     0022:
     0023:from collections import OrderedDict, defaultdict
     0024:from bb.utils import vercmp_string
     0025:
File: '/home/poky/bitbake/lib/bb/tinfoil.py', lineno: 18, function: <module>
     0014:import re
     0015:from collections import OrderedDict, defaultdict
     0016:
     0017:import bb.cache
 *** 0018:import bb.cooker
     0019:import bb.providers
     0020:import bb.taskdata
     0021:import bb.utils
     0022:import bb.command
File: '/home/poky/bitbake/lib/bb/cooker.py', lineno: 32, function: <module>
     0028:import queue
     0029:import signal
     0030:import subprocess
     0031:import errno
 *** 0032:import prserv.serv
     0033:import pyinotify
     0034:import json
     0035:import pickle
     0036:import codecs
File: '/home/poky/bitbake/lib/prserv/serv.py', lineno: 7, function: <module>
     0003:#
     0004:
     0005:import os,sys,logging
     0006:import signal, time
 *** 0007:from xmlrpc.server import SimpleXMLRPCServer, SimpleXMLRPCRequestHandler
     0008:import threading
     0009:import queue
     0010:import socket
     0011:import io
File: '/usr/lib/python3.8/xmlrpc/server.py', lineno: 117, function: <module>
     0113:import socketserver
     0114:import sys
     0115:import os
     0116:import re
 *** 0117:import pydoc
     0118:import traceback
     0119:try:
     0120:    import fcntl
     0121:except ImportError:
File: '/usr/lib/python3.8/pydoc.py', lineno: 370, function: <module>
     0366:    return module
     0367:
     0368:# ---------------------------------------------------- formatter base class
     0369:
 *** 0370:class Doc:
     0371:
     0372:    PYTHONDOCS = os.environ.get("PYTHONDOCS",
     0373:                                "https://docs.python.org/%d.%d/library"
     0374:                                % sys.version_info[:2])
File: '/usr/lib/python3.8/pydoc.py', lineno: 400, function: Doc
     0396:        raise TypeError(message)
     0397:
     0398:    docmodule = docclass = docroutine = docother = docproperty = docdata = fail
     0399:
 *** 0400:    def getdocloc(self, object, basedir=sysconfig.get_path('stdlib')):
     0401:        """Return the location of module docs or None"""
     0402:
     0403:        try:
     0404:            file = inspect.getabsfile(object)
File: '/usr/lib/python3.8/sysconfig.py', lineno: 512, function: get_path
     0508:    """Return a path corresponding to the scheme.
     0509:
     0510:    ``scheme`` is the install scheme name.
     0511:    """
 *** 0512:    return get_paths(scheme, vars, expand)[name]
     0513:
     0514:
     0515:def get_config_vars(*args):
     0516:    """With no arguments, return a dictionary of all configuration
File: '/usr/lib/python3.8/sysconfig.py', lineno: 502, function: get_paths
     0498:    ``scheme`` is the install scheme name. If not provided, it will
     0499:    return the default scheme for the current platform.
     0500:    """
     0501:    if expand:
 *** 0502:        return _expand_vars(scheme, vars)
     0503:    else:
     0504:        return _INSTALL_SCHEMES[scheme]
     0505:
     0506:
File: '/usr/lib/python3.8/sysconfig.py', lineno: 172, function: _expand_vars
     0168:def _expand_vars(scheme, vars):
     0169:    res = {}
     0170:    if vars is None:
     0171:        vars = {}
 *** 0172:    _extend_dict(vars, get_config_vars())
     0173:
     0174:    for key, value in _INSTALL_SCHEMES[scheme].items():
     0175:        if os.name in ('posix', 'nt'):
     0176:            value = os.path.expanduser(value)
File: '/usr/lib/python3.8/sysconfig.py', lineno: 550, function: get_config_vars
     0546:
     0547:        if os.name == 'nt':
     0548:            _init_non_posix(_CONFIG_VARS)
     0549:        if os.name == 'posix':
 *** 0550:            _init_posix(_CONFIG_VARS)
     0551:        # For backward compatibility, see issue19555
     0552:        SO = _CONFIG_VARS.get('EXT_SUFFIX')
     0553:        if SO is not None:
     0554:            _CONFIG_VARS['SO'] = SO
File: '/usr/lib/python3.8/sysconfig.py', lineno: 421, function: _init_posix
     0417:def _init_posix(vars):
     0418:    """Initialize the module as appropriate for POSIX systems."""
     0419:    # _sysconfigdata is generated at build time, see _generate_posix_vars()
     0420:    name = _get_sysconfigdata_name()
 *** 0421:    _temp = __import__(name, globals(), locals(), ['build_time_vars'], 0)
     0422:    build_time_vars = _temp.build_time_vars
     0423:    vars.update(build_time_vars)
     0424:
     0425:def _init_non_posix(vars):
Exception: ModuleNotFoundError: No module named '_sysconfigdata'

ERROR: libinput-1.12.6-r0 do_vigiles_pkg: No module named '_sysconfigdata'
ERROR: libinput-1.12.6-r0 do_vigiles_pkg: Function failed: do_tsmeta_pkgvars
ERROR: Logfile of failure stored in: /home/build/tmp/work/armv7vet2hf-neon-poky-linux-gnueabi/libinput/1.12.6-r0/temp/log.do_vigiles_pkg.13128
ERROR: Task (/home/poky/meta/recipes-graphics/wayland/libinput_1.12.6.bb:do_vigiles_pkg) failed with exit code '1'
ERROR: libpsl-0.20.2-r0 do_vigiles_pkg: Error executing a python function in exec_python_func() autogenerated:

The stack trace of python calls that resulted in this exception/failure was:
File: 'exec_python_func() autogenerated', lineno: 2, function: <module>
     0001:
 *** 0002:do_tsmeta_pkgvars(d)
     0003:
File: '/home/poky/meta-timesys/classes/tsmeta.bbclass', lineno: 466, function: do_tsmeta_pkgvars
     0462:
     0463:
     0464:python do_tsmeta_pkgvars() {
     0465:    tsmeta_get_pn(d)
 *** 0466:    tsmeta_get_src(d)
     0467:    tsmeta_get_pkg(d)
     0468:    tsmeta_get_packageconfig(d)
     0469:}
     0470:
File: '/home/poky/meta-timesys/classes/tsmeta.bbclass', lineno: 302, function: tsmeta_get_src
     0298:    return cve_v
     0299:
     0300:
     0301:def tsmeta_get_src(d):
 *** 0302:    import oe.recipeutils as oe
     0303:
     0304:    tsm_type = "src"
     0305:    src_dict = dict()
     0306:
File: '/home/poky/meta/lib/oe/recipeutils.py', lineno: 21, function: <module>
     0017:import shutil
     0018:import re
     0019:import fnmatch
     0020:import glob
 *** 0021:import bb.tinfoil
     0022:
     0023:from collections import OrderedDict, defaultdict
     0024:from bb.utils import vercmp_string
     0025:
File: '/home/poky/bitbake/lib/bb/tinfoil.py', lineno: 18, function: <module>
     0014:import re
     0015:from collections import OrderedDict, defaultdict
     0016:
     0017:import bb.cache
 *** 0018:import bb.cooker
     0019:import bb.providers
     0020:import bb.taskdata
     0021:import bb.utils
     0022:import bb.command
File: '/home/poky/bitbake/lib/bb/cooker.py', lineno: 32, function: <module>
     0028:import queue
     0029:import signal
     0030:import subprocess
     0031:import errno
 *** 0032:import prserv.serv
     0033:import pyinotify
     0034:import json
     0035:import pickle
     0036:import codecs
File: '/home/poky/bitbake/lib/prserv/serv.py', lineno: 7, function: <module>
     0003:#
     0004:
     0005:import os,sys,logging
     0006:import signal, time
 *** 0007:from xmlrpc.server import SimpleXMLRPCServer, SimpleXMLRPCRequestHandler
     0008:import threading
     0009:import queue
     0010:import socket
     0011:import io
File: '/usr/lib/python3.8/xmlrpc/server.py', lineno: 117, function: <module>
     0113:import socketserver
     0114:import sys
     0115:import os
     0116:import re
 *** 0117:import pydoc
     0118:import traceback
     0119:try:
     0120:    import fcntl
     0121:except ImportError:
File: '/usr/lib/python3.8/pydoc.py', lineno: 370, function: <module>
     0366:    return module
     0367:
     0368:# ---------------------------------------------------- formatter base class
     0369:
 *** 0370:class Doc:
     0371:
     0372:    PYTHONDOCS = os.environ.get("PYTHONDOCS",
     0373:                                "https://docs.python.org/%d.%d/library"
     0374:                                % sys.version_info[:2])
File: '/usr/lib/python3.8/pydoc.py', lineno: 400, function: Doc
     0396:        raise TypeError(message)
     0397:
     0398:    docmodule = docclass = docroutine = docother = docproperty = docdata = fail
     0399:
 *** 0400:    def getdocloc(self, object, basedir=sysconfig.get_path('stdlib')):
     0401:        """Return the location of module docs or None"""
     0402:
     0403:        try:
     0404:            file = inspect.getabsfile(object)
File: '/usr/lib/python3.8/sysconfig.py', lineno: 512, function: get_path
     0508:    """Return a path corresponding to the scheme.
     0509:
     0510:    ``scheme`` is the install scheme name.
     0511:    """
 *** 0512:    return get_paths(scheme, vars, expand)[name]
     0513:
     0514:
     0515:def get_config_vars(*args):
     0516:    """With no arguments, return a dictionary of all configuration
File: '/usr/lib/python3.8/sysconfig.py', lineno: 502, function: get_paths
     0498:    ``scheme`` is the install scheme name. If not provided, it will
     0499:    return the default scheme for the current platform.
     0500:    """
     0501:    if expand:
 *** 0502:        return _expand_vars(scheme, vars)
     0503:    else:
     0504:        return _INSTALL_SCHEMES[scheme]
     0505:
     0506:
File: '/usr/lib/python3.8/sysconfig.py', lineno: 172, function: _expand_vars
     0168:def _expand_vars(scheme, vars):
     0169:    res = {}
     0170:    if vars is None:
     0171:        vars = {}
 *** 0172:    _extend_dict(vars, get_config_vars())
     0173:
     0174:    for key, value in _INSTALL_SCHEMES[scheme].items():
     0175:        if os.name in ('posix', 'nt'):
     0176:            value = os.path.expanduser(value)
File: '/usr/lib/python3.8/sysconfig.py', lineno: 550, function: get_config_vars
     0546:
     0547:        if os.name == 'nt':
     0548:            _init_non_posix(_CONFIG_VARS)
     0549:        if os.name == 'posix':
 *** 0550:            _init_posix(_CONFIG_VARS)
     0551:        # For backward compatibility, see issue19555
     0552:        SO = _CONFIG_VARS.get('EXT_SUFFIX')
     0553:        if SO is not None:
     0554:            _CONFIG_VARS['SO'] = SO
File: '/usr/lib/python3.8/sysconfig.py', lineno: 421, function: _init_posix
     0417:def _init_posix(vars):
     0418:    """Initialize the module as appropriate for POSIX systems."""
     0419:    # _sysconfigdata is generated at build time, see _generate_posix_vars()
     0420:    name = _get_sysconfigdata_name()
 *** 0421:    _temp = __import__(name, globals(), locals(), ['build_time_vars'], 0)
     0422:    build_time_vars = _temp.build_time_vars
     0423:    vars.update(build_time_vars)
     0424:
     0425:def _init_non_posix(vars):
Exception: ModuleNotFoundError: No module named '_sysconfigdata'

ERROR: libpsl-0.20.2-r0 do_vigiles_pkg: No module named '_sysconfigdata'
ERROR: libpsl-0.20.2-r0 do_vigiles_pkg: Function failed: do_tsmeta_pkgvars
ERROR: Logfile of failure stored in: /home/build/tmp/work/armv7vet2hf-neon-poky-linux-gnueabi/libpsl/0.20.2-r0/temp/log.do_vigiles_pkg.13129
ERROR: Task (/home/poky/meta/recipes-support/libpsl/libpsl_0.20.2.bb:do_vigiles_pkg) failed with exit code '1'

```

**I have installed `Python 3.6.9` and set that as my default Python3 version, that resolved the all issues and the Vigiles Manifest got generated.** 

Even though am aware that it could be due to the fact that am using a Host distro that was not validated by Yocto. I thought of raising this issue, since all other packages were built fine and if there is something that you can patch in this layer to avoid this error. 

If it is already in your radar, and is being addressed. Please close this post. Thanks.

**Comment by nodeax** [comment]
Thanks for reporting this issue and for providing detailed logs :) 
The issue has been resolved with ad7a55d0c6050bbef036510c167aff9168dd9bd6.

**Comment by Ole2mail** [comment]
got the same issue with the command on python-3.8.5

devtool modify u-boot 

`ERROR: Error executing a python function in exec_python_func() autogenerated:

The stack trace of python calls that resulted in this exception/failure was:
File: 'exec_python_func() autogenerated', lineno: 2, function: <module>
     0001:
 *** 0002:devtool_post_unpack(d)
     0003:
File: '/media/compilation/projects/pinebookpro/poky/meta/classes/devtool-source.bbclass', lineno: 68, function: devtool_post_unpack
     0064:}
     0065:
     0066:
     0067:python devtool_post_unpack() {
 *** 0068:    import oe.recipeutils
     0069:    import shutil
     0070:    sys.path.insert(0, os.path.join(d.getVar('COREBASE'), 'scripts', 'lib'))
     0071:    import scriptutils
     0072:    from devtool import setup_git_repo
File: '/media/compilation/projects/pinebookpro/poky/meta/lib/oe/recipeutils.py', lineno: 21, function: <module>
     0017:import shutil
     0018:import re
     0019:import fnmatch
     0020:import glob
 *** 0021:import bb.tinfoil
     0022:
     0023:from collections import OrderedDict, defaultdict
     0024:from bb.utils import vercmp_string
     0025:
File: '/media/compilation/projects/pinebookpro/poky/bitbake/lib/bb/tinfoil.py', lineno: 19, function: <module>
     0015:from collections import OrderedDict, defaultdict
     0016:from functools import partial
     0017:
     0018:import bb.cache
 *** 0019:import bb.cooker
     0020:import bb.providers
     0021:import bb.taskdata
     0022:import bb.utils
     0023:import bb.command
File: '/media/compilation/projects/pinebookpro/poky/bitbake/lib/bb/cooker.py', lineno: 25, function: <module>
     0021:import bb, bb.exceptions, bb.command
     0022:from bb import utils, data, parse, event, cache, providers, taskdata, runqueue, build
     0023:import queue
     0024:import signal
 *** 0025:import prserv.serv
     0026:import pyinotify
     0027:import json
     0028:import pickle
     0029:import codecs
File: '/media/compilation/projects/pinebookpro/poky/bitbake/lib/prserv/serv.py', lineno: 7, function: <module>
     0003:#
     0004:
     0005:import os,sys,logging
     0006:import signal, time
 *** 0007:from xmlrpc.server import SimpleXMLRPCServer, SimpleXMLRPCRequestHandler
     0008:import threading
     0009:import queue
     0010:import socket
     0011:import io
File: '/usr/lib/python3.8/xmlrpc/server.py', lineno: 117, function: <module>
     0113:import socketserver
     0114:import sys
     0115:import os
     0116:import re
 *** 0117:import pydoc
     0118:import traceback
     0119:try:
     0120:    import fcntl
     0121:except ImportError:
File: '/usr/lib/python3.8/pydoc.py', lineno: 370, function: <module>
     0366:    return module
     0367:
     0368:# ---------------------------------------------------- formatter base class
     0369:
 *** 0370:class Doc:
     0371:
     0372:    PYTHONDOCS = os.environ.get("PYTHONDOCS",
     0373:                                "https://docs.python.org/%d.%d/library"
     0374:                                % sys.version_info[:2])
File: '/usr/lib/python3.8/pydoc.py', lineno: 400, function: Doc
     0396:        raise TypeError(message)
     0397:
     0398:    docmodule = docclass = docroutine = docother = docproperty = docdata = fail
     0399:
 *** 0400:    def getdocloc(self, object, basedir=sysconfig.get_path('stdlib')):
     0401:        """Return the location of module docs or None"""
     0402:
     0403:        try:
     0404:            file = inspect.getabsfile(object)
File: '/usr/lib/python3.8/sysconfig.py', lineno: 512, function: get_path
     0508:    """Return a path corresponding to the scheme.
     0509:
     0510:    ``scheme`` is the install scheme name.
     0511:    """
 *** 0512:    return get_paths(scheme, vars, expand)[name]
     0513:
     0514:
     0515:def get_config_vars(*args):
     0516:    """With no arguments, return a dictionary of all configuration
File: '/usr/lib/python3.8/sysconfig.py', lineno: 502, function: get_paths
     0498:    ``scheme`` is the install scheme name. If not provided, it will
     0499:    return the default scheme for the current platform.
     0500:    """
     0501:    if expand:
 *** 0502:        return _expand_vars(scheme, vars)
     0503:    else:
     0504:        return _INSTALL_SCHEMES[scheme]
     0505:
     0506:
File: '/usr/lib/python3.8/sysconfig.py', lineno: 172, function: _expand_vars
     0168:def _expand_vars(scheme, vars):
     0169:    res = {}
     0170:    if vars is None:
     0171:        vars = {}
 *** 0172:    _extend_dict(vars, get_config_vars())
     0173:
     0174:    for key, value in _INSTALL_SCHEMES[scheme].items():
     0175:        if os.name in ('posix', 'nt'):
     0176:            value = os.path.expanduser(value)
File: '/usr/lib/python3.8/sysconfig.py', lineno: 550, function: get_config_vars
     0546:
     0547:        if os.name == 'nt':
     0548:            _init_non_posix(_CONFIG_VARS)
     0549:        if os.name == 'posix':
 *** 0550:            _init_posix(_CONFIG_VARS)
     0551:        # For backward compatibility, see issue19555
     0552:        SO = _CONFIG_VARS.get('EXT_SUFFIX')
     0553:        if SO is not None:
     0554:            _CONFIG_VARS['SO'] = SO
File: '/usr/lib/python3.8/sysconfig.py', lineno: 421, function: _init_posix
     0417:def _init_posix(vars):
     0418:    """Initialize the module as appropriate for POSIX systems."""
     0419:    # _sysconfigdata is generated at build time, see _generate_posix_vars()
     0420:    name = _get_sysconfigdata_name()
 *** 0421:    _temp = __import__(name, globals(), locals(), ['build_time_vars'], 0)
     0422:    build_time_vars = _temp.build_time_vars
     0423:    vars.update(build_time_vars)
     0424:
     0425:def _init_non_posix(vars):
Exception: ModuleNotFoundError: No module named '_sysconfigdata'
`
